Add instrumentation API and native OpenTelemetry implementation

AlexanderWert opened the pull request:
Adds an instrumentation API with request & response lifecycle callbacks, and an implementation based on OpenTelemetry that is used by default.

To be able to implement the [OpenTelemetry semantic conventions](https://opentelemetry.io/docs/specs/semconv/database/elasticsearch/#url-path-parts), this PR also adds a `pathParameters` method to `Endpoint` that provides access to the values in URL template placeholders.

swallez reviewed (commented):
Thanks! Overall it looks nice.

Rather than the route (i.e. path template), using the API endpoint identifier would make things easier. This identifier comes from the API specification (see e.g. [IndexRequest](https://github.com/elastic/elasticsearch-specification/blob/7ddbec9091a8aa2c7423e2df14f1c3394f421627/specification/_global/index/IndexRequest.ts#L36)) and is available as `endpoint.id()`.

Benefits:
* a single endpoint can have several routes and different http methods depending on which optional path parameters are set,
* using endpoint id avoids doing some pattern matching on the route for things like document id extraction.
* naming the span with the endpoint id ensures a 1:1 mapping from APIs to span names regardless of the path parameters set.


Note: I suggested this to @estolfo in a recent discussion.

AlexanderWert commented on java-client/src/main/java/co/elastic/clients/transport/Endpoint.java:
```diff
@@ -56,6 +56,18 @@
+  /**
+   * Get the route for a request (i.e. URL pattern).
+   */
+  String route(RequestT request);
```
agreed on using the endpoint identifier for the span name.

AlexanderWert commented on java-client/src/main/java/co/elastic/clients/transport/rest_client/InstrumentationUtil.java:
```diff
@@ -0,0 +1,131 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class InstrumentationUtil {
```
done

AlexanderWert commented on java-client/src/main/java/co/elastic/clients/transport/rest_client/InstrumentationUtil.java:
```diff
@@ -0,0 +1,131 @@
+            Map<String, String> queryParameters = endpoint.queryParameters(request);
+            if (!queryParameters.isEmpty()) {
+                String queryString =
+                        queryParameters.entrySet().stream().map(e -> e.getKey() + "=" + e.getValue()).collect(Collectors.joining("&"));
```
not relevant anymore

AlexanderWert commented on java-client/src/main/java/co/elastic/clients/transport/rest_client/InstrumentationUtil.java:
```diff
@@ -0,0 +1,131 @@
+        String route = endpoint.route(request);
+
+        // We capture the request body in the span only for search-type requests.
+        return route.contains("/_search") ||
```
done.

AlexanderWert commented on java-client/src/main/java/co/elastic/clients/transport/rest_client/InstrumentationUtil.java:
```diff
@@ -0,0 +1,131 @@
+        span.setAttribute("url.scheme", host.getSchemeName());
+    }
+
+    protected <RequestT> void captureBody(@Nullable Span span, RequestT request, Endpoint<RequestT, ?, ?> endpoint,
```
done

AlexanderWert commented on java-client/build.gradle.kts:
```diff
@@ -201,6 +202,13 @@ dependencies {
+    // Apache 2.0
+    // https://github.com/open-telemetry/opentelemetry-java
+    implementation("io.opentelemetry", "opentelemetry-api", openTelemetryVersion)
+    implementation("io.opentelemetry", "opentelemetry-semconv", "$openTelemetryVersion-alpha")
```
This is only used for OTel attribute names.

AlexanderWert commented:
@swallez 

As https://github.com/open-telemetry/semantic-conventions/pull/23 has been merged, I updated this PR accordingly and also incorporated your suggestions.
I think it's safe now to finalize this PR.

swallez reviewed (approved):
LGTM
